src/presentation/api/controller/services.go

```diff
@@ -0,0 +1,253 @@
+	persistentDbService *internalDbInfra.PersistentDatabaseService,
+) *ServicesController {
+	return &ServicesController{
+		servicesService: service.NewServicesService(persistentDbService),
+	}
+}
+
+// ReadServices	 godoc
+// @Summary      ReadServices
+// @Description  List installed services and their status.
+// @Tags         services
+// @Security     Bearer
+// @Accept       json
+// @Produce      json
+// @Success      200 {array} dto.InstalledServiceWithMetrics
+// @Router       /v1/services/ [get]
+func (controller *ServicesController) Read(c echo.Context) error {
+	return apiHelper.ServiceResponseWrapper(c, controller.servicesService.Read())
+}
+
+// ReadInstallableServices	 godoc
+// @Summary      ReadInstallableServices
+// @Description  List installable services.
+// @Tags         services
+// @Security     Bearer
+// @Accept       json
+// @Produce      json
+// @Success      200 {array} entity.InstallableService
+// @Router       /v1/services/installables/ [get]
+func (controller *ServicesController) ReadInstallables(c echo.Context) error {
+	return apiHelper.ServiceResponseWrapper(
+		c, controller.servicesService.ReadInstallables(),
+	)
+}
+
+func parseRawEnvs(envs interface{}) ([]string, error) {
+	rawEnvs := []string{}
+	rawEnvsSlice, assertOk := envs.([]interface{})
+	if !assertOk {
+		rawEnvUnique, assertOk := envs.(string)
+		if !assertOk {
+			return rawEnvs, errors.New("EnvsMustBeStringOrStringSlice")
+		}
+		rawEnvsSlice = []interface{}{rawEnvUnique}
+	}
+
+	for _, rawEnv := range rawEnvsSlice {
+		rawEnvStr, err := voHelper.InterfaceToString(rawEnv)
+		if err != nil {
+			slog.Debug(err.Error(), slog.Any("env", rawEnv))
+			continue
+		}
+		rawEnvs = append(rawEnvs, rawEnvStr)
+	}
+
+	return rawEnvs, nil
+}
+
+func parseRawPortBindings(bindings interface{}) ([]string, error) {
+	rawPortBindings := []string{}
+	rawPortBindingsSlice, assertOk := bindings.([]interface{})
+	if !assertOk {
+		rawPortBindingUnique, assertOk := bindings.(map[string]interface{})
+		if !assertOk {
+			return rawPortBindings, errors.New("PortBindingsMustBeMapOrMapSlice")
+		}
+		rawPortBindingsSlice = []interface{}{rawPortBindingUnique}
+	}
+
+	for _, rawPortBinding := range rawPortBindingsSlice {
+		rawPortBindingMap, assertOk := rawPortBinding.(map[string]interface{})
+		if !assertOk {
+			slog.Debug(
+				"InvalidPortBindingStructure", slog.Any("portBinding", rawPortBinding),
+			)
+			continue
+		}
+
+		rawPortStr, err := voHelper.InterfaceToString(rawPortBindingMap["port"])
+		if err != nil {
+			slog.Debug(err.Error(), slog.String("port", rawPortStr))
+			continue
+		}
+
+		rawProtocolStr := ""
+		if _, protocolInputExists := rawPortBindingMap["protocol"]; protocolInputExists {
+			rawProtocolStr, err = voHelper.InterfaceToString(rawPortBindingMap["protocol"])
+			if err != nil {
+				slog.Debug(err.Error(), slog.String("port", rawPortStr))
+				continue
+			}
+		}
+
+		rawPortBindingStr := rawPortStr + "/" + rawProtocolStr
```

Comment:
You should init rawPortBindingStr before checking protocol. Then inside the if of the protocol you add "/" + rawProtocolStr to rawPortBindingStr. That way you avoid sending a port binding with a blank / in case the protocol is never sent. You may be able to ditch the loose / at the end, but it's best to send in the correct format.

Reply:
Done!